surveys/promise-rejections/survey.md

```diff
@@ -0,0 +1,152 @@
+# Node.js Promise reject use case survey
+
+Today, Node.js handles unhandled rejections by emitting a deprecation warning to stderr. The warning shows the stack where the rejection happened, and states that in future Node.js versions unhandled rejections will result on Node.js exitting with non-zero status code. We intend to remove the deprecation warning, replacing it with a stable behavior which might be different from the one described on the deprecation warning. We're running this survey to better understand how Node.js users are using Promises and how they are dealing with unhandled rejections today, so we can make an informed decision on how to move forward.
+
+## What is a Promise rejection?
+
+A Promise rejection indicates that something went wrong while executing a Promise or an async function. Rejections can occur on several situations: thorwing inside an async function or a `Promise` executor/then/catch/finally callback, when calling the `reject` callback of an `executor`, or when calling `Promise.reject`.
+
+```js
+
+Promise.reject()  // This will result in a rejection
+
+new Promise((fulfill, reject) => {
+  reject();  // This will result in a rejection
+});
+
+new Promise(() => {
+  throw new Error();  // This will result in a rejection
+}).then(() => {
+  throw new Error();  // This will result in a rejection
+}, () => {
+  throw new Error();  // This will result in a rejection
+}).catch(() => {
+  throw new Error();  // This will result in a rejection
+}).finally(() => {
+  throw new Error();  // This will result in a rejection
+});
+
+async function () {
+  throw new Error();  // This will result in a rejection
+}
+
+async function () {
+  if (a === undefined)
+    a();  // This will result in a rejection
+}
+```
+
+## What is an unhandled rejection?
+
+Rejections are designed to be handled asynchronously. In JavaScript, Promises can be handled when awaiting for a promise inside a try/catch block, or via .catch calls:
+
+
+```js
+async function foo() {
+  throw new Error();
+}
+
+foo().catch(() => console.error("an error occured"));  // 1. Handled
+
+try {
+  await foo();  // 2. Handled
+} catch(e) {
+  console.error("an error occured")
+}
+
+foo(); // 3. Unhandled, but execution continues
+
+const rejected = foo(); // 4. Unhandled on current event loop tick
+
+// 4. ... but handled asynchronously when the setTimeout callback is popped from
+// the event loop
+setTimeout(() => rejected.catch(() => console.error("an error occured")), 100);
+```
+
+An unhandled rejection is a rejection which hasn't been handled yet. It might be handled in the future, like example 4, but it might also stay unhandled forever (like example 3).
+
```

Comment:
Suggestion. Add the following paragraph, so that users are aware of potential problems caused by unhandled rejections.:
```
Unhandled rejections may leave your application in a non-deterministic state, whether it's internal application state (including memory leaks), external resources used by your application (say, file handles or database connections), or external state (say, consistency of data in a database).
```

I don't mind changing the text in a more readable/concise manner.

Reply:
I think that's a bit scary and Jordan and Gus would probably object to that phrasing. I suggest alternatively:

> Certain unhandled rejections may in rare cases leave your application in a non-deterministic and unsafe state, whether it's internal application state (including memory leaks), external resources used by your application (say, file handles or database connections), or external state (say, consistency of data in a database).

I think it's important to communicate that it's not normal for any unhandled rejection to leave your app in inconsistent state but that it's possible.

Reply:
I'm fine with the proposed change.